### PR TITLE
feat(workers/branch): allow to define a blocked label

### LINKED ISSUE #6552: Allow to define a 'blocked' label
**What would you like Renovate to be able to do?**

If there is a certain label on a renovate PR not more rebases will happen.
There are issues (mainly major updates) you cannot do, because that would lead to breaking changes for your users. We currently flag them with a label to come back later and resolve this for the next major release.

While these PRs are blocked renovate does rebase them again and again which wouldn't be necessary.

**Describe the solution you'd like**

It would be nice to configure a label from which renovate can detect to ignore this PR.
It shouldn't be closed nor touched otherwise.

### PR DESCRIPTION
**Pull request by olegkrivtsov**
## Changes:

This PR is work in progress for #6552. This issue is rather complex, so I decided to push what I have for now and request for a brief review/advice.

Could you please review my changes and tell me if I understand the task correctly and move in the right direction? In particular, it is not clear for now where the branch rebase actually happens. I guess it happens in the `tryBranchAutomerge()` function:
https://github.com/renovatebot/renovate/blob/7e069f4adce758bef94fc31ea4db8984f3169da9/lib/workers/branch/index.ts#L483

If that's correct, how do I determine if the branch was actually rebased, not merged? 

Once it is more clear for me, I'll proceed with creating unit tests for this and testing agains a real repo.

## Context:

Closes #6552 

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by olegkrivtsov** [comment]
Hi @rarkins thank you for the info. After reading https://docs.renovatebot.com/updating-rebasing/ I've realized that Renovate doesn't use the "standard" git rebase, instead it just recreates the branch and pushes one commit into it. So I improved my changes and committed what I have right now: https://github.com/renovatebot/renovate/pull/12164/commits/9664a2eb4b8375af083500a020ffaa980dbb0bd4

I'm currently trying to test if my changes really work on my real repo: https://github.com/olegkrivtsov/renovate-stop-rebasing-test/blob/master/docker-compose.yml My `docker-compose.yaml` file contains one dependency `httpd:1.0.0` 

1. I ran `yarn start` to cause the Renovate's branch/PR creation: https://github.com/olegkrivtsov/renovate-stop-rebasing-test/pull/10
2. To cause it rebase/recreation of the Renovate's branch, I edited the dependency `httpd:1.0.1` and ran Renovate again
3. For some reason I don't see it really rebased/recreated the branch

So, I'd like to ask you, what is the correct way to test the rebase of a branch?

**Comment by rarkins** [comment]
Easiest thing you can do is create a package.json with two out of date dependencies, then you get two PRs, then you merge one of them. You should be left with one conflicted PR which Renovate would normally rebase to de-conflict it. If instead you apply the label, it should ignore it.

Another alternative is to configure rebaseWhen=behind-base-branch and have a single PR. Then you just make commits to the main branch (e.g. README updates) and you should see Renovate rebases the PR each time. Now apply the label and observe that it stops rebasing.

Both the above use cases could be good to test against

File: docs/usage/configuration-options.md [hunk 1]
```diff
@@ -2412,6 +2412,11 @@ This works because Renovate will add a "renovate/stability-days" pending status
 
 <!-- markdownlint-enable MD001 -->
 
+## stopRebasingLabel
+
+On supported platforms it is possible to add a label to a PR to request Renovate stop rebasing the PR branch after the first rebase.
```
Comment (rarkins) [inline docs/usage/configuration-options.md:2417]:
```suggestion
On supported platforms it is possible to add a label to a PR to request Renovate stop rebasing the PR.
```

File: lib/config/options/index.ts [hunk 2]
```diff
@@ -1228,6 +1228,12 @@ const options: RenovateOptions[] = [
     type: 'string',
     default: 'rebase',
   },
+  {
+    name: 'stopRebasingLabel',
```
Comment (rarkins) [inline lib/config/options/index.ts:1232]: Add `supportedPlatforms` so this

File: lib/workers/branch/index.ts [hunk 3]
```diff
@@ -441,6 +441,32 @@ export async function processBranch(
...
+        return {
+          branchExists: true,
+          prNo: branchPr?.number,
+          result: BranchResult.NoWork,
+        };
+      }
+
+      logger.debug(
+        'Branch will be rebased if necessary despite of stopRebasingLabel since Rebase checkbox is checked in PR body'
+      );
```
Comment (rarkins) [inline lib/workers/branch/index.ts:467]:
```suggestion
```

File: lib/workers/branch/index.ts [hunk 4]
```diff
@@ -462,6 +488,38 @@ export async function processBranch(
...
+        number: branchPr.number,
+        prTitle: branchPr.title,
+        prBody: newBody,
+        platformOptions: getPlatformPrOptions(config),
+      });
+
+      // This flag may be used by ensurePr to generate the correct PR body
+      config.prRebaseBoxUnchecked = true;
+    }
+
```
Comment (rarkins) [inline lib/workers/branch/index.ts:522]: This is too hacky of a way to do it. It should not need its own `platform.updatePr()` and instead should leverage the existing logic

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/workers/branch/index.ts [hunk 5]
```diff
@@ -440,6 +440,28 @@ export async function processBranch(
     }
     const forcedManually = userRebaseRequested || !branchExists;
     config.forceCommit = forcedManually || branchPr?.isConflicted;
+
+    const stopRebasingLabelPresents = branchPr?.labels?.includes(
```
Comment (rarkins) [inline lib/workers/branch/index.ts:444]:
```suggestion
    const stopRebasingLabelPresent = branchPr?.labels?.includes(
```

File: lib/workers/pr/body/controls.ts [hunk 6]
```diff
@@ -6,5 +6,13 @@ export async function getControls(config: BranchConfig): Promise<string> {
   const warning = (await isBranchModified(config.branchName))
     ? emojify(' :warning: **Warning**: custom changes will be lost.')
     : '';
-  return `\n\n---\n\n - [ ] <!-- rebase-check -->If you want to rebase/retry this PR, click this checkbox.${warning}\n\n`;
+
+  const rebaseBoxUncheckedWarning =
+    config?.prRebaseBoxUnchecked === true
+      ? emojify(
+          '\n\n :memo: **Note**: The rebase/retry checkbox was unchecked because of `stopRebasingLabel` setting in your configuration'
```
Comment (rarkins) [inline lib/workers/pr/body/controls.ts:13]: No, we shouldn't do this. If someone has the stop-rebasing label AND ticks the checkbox to rebase, we should still rebase.

Comment (rarkins) [reply lib/workers/pr/body/controls.ts:13]: Are you sure any of these changes are necessary?

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/workers/pr/body/controls.ts (hunk 6 above)
Comment (rarkins) [reply lib/workers/pr/body/controls.ts:13]: I think it's this text you need to change instead:

image

File: lib/workers/branch/index.ts [hunk 7]
```diff
@@ -440,6 +440,29 @@ export async function processBranch(
...
+      config.stopRebasingLabel
+    );
+    config.stopRebasingLabelPresent = stopRebasingLabelPresent;
+
+    const prRebaseChecked = branchPr?.body?.includes(
+      `- [x] <!-- rebase-check -->`
+    );
+
+    if (branchExists && stopRebasingLabelPresent) {
+      if (!prRebaseChecked) {
```
Comment (rarkins) [inline lib/workers/branch/index.ts:454]: We might also need to check for dashboard check

File: lib/workers/pr/body/config-description.ts [hunk 8]
```diff
@@ -42,7 +42,9 @@ export async function getPrConfigDescription(
   }
   prBody += '\n\n';
   prBody += emojify(':recycle: **Rebasing**: ');
-  if (config.rebaseWhen === 'behind-base-branch') {
+  if (config.stopRebasingLabelPresent) {
+    prBody += 'Only once (due to `stopRebasingLabel` config option)';
+  } else if (config.rebaseWhen === 'behind-base-branch') {
```
Comment (rarkins) [inline lib/workers/pr/body/config-description.ts:47]: Revert these changes and instead just combine it with rebaseWhen===never below

File: lib/workers/pr/body/config-description.ts [hunk 9]
```diff
@@ -42,7 +42,9 @@ export async function getPrConfigDescription(
...
     prBody += 'Whenever PR is behind base branch';
   } else if (config.rebaseWhen === 'never') {
```
Comment (rarkins) [inline lib/workers/pr/body/config-description.ts:49]:
```suggestion
  } else if (config.rebaseWhen === 'never' || config.stopRebasingLabelPresent) {
```

File: lib/workers/pr/body/controls.ts [hunk 10]
```diff
@@ -6,5 +6,6 @@ export async function getControls(config: BranchConfig): Promise<string> {
   const warning = (await isBranchModified(config.branchName))
     ? emojify(' :warning: **Warning**: custom changes will be lost.')
     : '';
+
```
Comment (rarkins) [inline lib/workers/pr/body/controls.ts:9]:
```suggestion
```

**Review by rarkins — CHANGES_REQUESTED** [review]
I would also like to change the term here from "rebasing" to instead "updating", e.g. `stopUpdatingLabel`, defaulting to `stop-updating`)

File: docs/usage/configuration-options.md [hunk 11]
```diff
@@ -2523,6 +2523,11 @@ This works because Renovate will add a "renovate/stability-days" pending status
 
 <!-- markdownlint-enable MD001 -->
 
+## stopRebasingLabel
```
Comment (rarkins) [inline docs/usage/configuration-options.md:2526]:
```suggestion
## stopUpdatingLabel
```

File: docs/usage/configuration-options.md [hunk 12]
```diff
@@ -2523,6 +2523,11 @@ This works because Renovate will add a "renovate/stability-days" pending status
...
+
+On supported platforms it is possible to add a label to a PR to request Renovate stop rebasing the PR.
+By default this label is `"stop-rebasing"` however you can configure it to anything you want by changing this `stopRebasingLabel` field.
```
Comment (rarkins) [inline docs/usage/configuration-options.md:2529]:
```suggestion
On supported platforms it is possible to add a label to a PR to request Renovate stop updating the PR.
By default this label is `"stop-updating"` however you can configure it to anything you want by changing this `stopUpdatingLabel` field.
```

File: lib/config/options/index.ts [hunk 13]
```diff
@@ -1305,6 +1305,13 @@ const options: RenovateOptions[] = [
...
```
Comment (rarkins) [inline lib/config/options/index.ts:1309]:
```suggestion
    name: 'stopUpdatingLabel',
```

File: lib/config/options/index.ts [hunk 14]
```diff
@@ -1305,6 +1305,13 @@ const options: RenovateOptions[] = [
...
+    description: 'Label to use to request the bot to stop rebasing a PR.',
```
Comment (rarkins) [inline lib/config/options/index.ts:1310]:
```suggestion
    description: 'Label to use to request the bot to stop updating a PR.',
```

File: lib/config/options/index.ts [hunk 15]
```diff
@@ -1305,6 +1305,13 @@ const options: RenovateOptions[] = [
...
+    type: 'string',
+    default: 'stop-rebasing',
```
Comment (rarkins) [inline lib/config/options/index.ts:1312]:
```suggestion
    default: 'stop-updating',
```

**Review by rarkins — CHANGES_REQUESTED** [review]
Please make other changes to naming accordingly

File: lib/config/options/index.ts [hunk 16]
```diff
@@ -1305,6 +1305,13 @@ const options: RenovateOptions[] = [
     type: 'string',
     default: 'rebase',
   },
+  {
+    name: 'stopUpdatingLabel',
+    description: 'Label to use to request the bot to stop updating a PR.',
+    type: 'string',
+    default: 'stop-updating',
+    supportedPlatforms: ['github'],
```
Comment (rarkins) [inline lib/config/options/index.ts:1313]: Shouldn't this work on any platform which supports PR labels and which returns them when we call `getPr()`?

File: lib/workers/branch/index.ts [hunk 17]
```diff
@@ -440,6 +440,29 @@ export async function processBranch(
     }
     const forcedManually = userRebaseRequested || !branchExists;
     config.forceCommit = forcedManually || branchPr?.isConflicted;
+
+    const stopUpdatingLabelPresent = branchPr?.labels?.includes(
+      config.stopUpdatingLabel
+    );
+    config.stopUpdatingLabelPresent = stopUpdatingLabelPresent;
```
Comment (rarkins) [inline lib/workers/branch/index.ts:447]:
```suggestion
    config.stopUpdatingLabelPresent = branchPr?.labels?.includes(
      config.stopUpdatingLabel
    );
```

File: lib/workers/branch/index.ts [hunk 18]
```diff
@@ -440,6 +440,29 @@ export async function processBranch(
...
+
+    const prRebaseChecked = branchPr?.body?.includes(
+      `- [x] <!-- rebase-check -->`
+    );
+
+    if (branchExists && dependencyDashboardCheck && stopUpdatingLabelPresent) {
```
Comment (rarkins) [inline lib/workers/branch/index.ts:453]:
```suggestion
    if (branchExists && dependencyDashboardCheck && config.stopUpdatingLabelPresent) {
```

This probably will now fail prettier

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/config/options/index.ts (hunk 16 above)
Comment (olegkrivtsov) [reply lib/config/options/index.ts]: Yes, that's right

File: lib/workers/types.ts [hunk 19]
```diff
@@ -115,4 +115,5 @@ export interface BranchConfig
   packageFiles?: Record<string, PackageFile[]>;
   prBlockedBy?: PrBlockedBy;
   prNo?: number;
+  stopUpdatingLabelPresent?: boolean;
```
Comment (rarkins) [inline lib/workers/types.ts:118]:
```suggestion
  stopUpdating?: boolean;
```

Let's shorten this (requires some other code changes too in other files)

**Review by rarkins — CHANGES_REQUESTED** [review]

**Review by rarkins — DISMISSED** [review]

**Review by rarkins — APPROVED** [review]

**Review by viceice — APPROVED** [review]
